File: game-server/src/main/java/com/kingston/mmorpg/game/ai/fsm/AiTest.java
package com.kingston.mmorpg.game.ai.fsm;

import com.kingston.mmorpg.game.database.user.entity.PlayerEnt;
import com.kingston.mmorpg.game.scene.actor.Monster;
import com.kingston.mmorpg.game.scene.model.Life;

public class AiTest {

	public static void main(String[] args) throws Exception {
		PlayerEnt player = new PlayerEnt();
		player.setLife(new Life(150, 150));
		Monster monster = new Monster();
		monster.setLife(new Life(120, 120));
		
		Scene scene = new Scene();
		scene.setPlayer(player);
		scene.setMonster(monster);

		player.setScene(scene);
		monster.setScene(scene);

		State patrolState = new PatrolState();
		State attackState = new AttackState();
		State runState = new RunAwayState();

		Transition transition1 = new Patrol2AttackTransition(patrolState, attackState);
		Transition transition2 = new Attack2RunTransition(attackState, runState);
		Transition transition3 = new Atttack2PatrolTransition(attackState, patrolState);
		Transition transition4 = new Run2PatrolTransition(runState, patrolState);

		FiniteStateMachine fsm = new FiniteStateMachine();
		fsm.setInitState(patrolState);

		fsm.addTransition(transition1);
		fsm.addTransition(transition2);
		fsm.addTransition(transition3);
		fsm.addTransition(transition4);

		while (true) {
			fsm.enterFrame(player);
			Thread.sleep(500);
		}

	}

}

File: game-server/src/main/java/org/forfun/mmorpg/game/ai/fsm/AiTest.kt
package org.forfun.mmorpg.game.ai.fsm

import org.forfun.mmorpg.game.database.user.entity.PlayerEnt
import org.forfun.mmorpg.game.scene.actor.Monster
import org.forfun.mmorpg.game.scene.model.Life

object AiTest {

    @JvmStatic
    fun main(args: Array<String>) {
        val player = PlayerEnt()
        player.life = Life(150, 150)
        val monster = Monster()
        monster.life = Life(120, 120)

        val scene = Scene()
        scene.player = player
        scene.monster = monster

        player.scene = scene
        monster.scene = scene

        val patrolState = PatrolState()
        val attackState = AttackState()
        val runState = RunAwayState()

        val transition1 = Patrol2AttackTransition(patrolState, attackState)
        val transition2 = Attack2RunTransition(attackState, runState)
        val transition3 = Atttack2PatrolTransition(attackState, patrolState)
        val transition4 = Run2PatrolTransition(runState, patrolState)

        val fsm = FiniteStateMachine()
        fsm.initState = patrolState

        fsm.addTransition(transition1)
        fsm.addTransition(transition2)
        fsm.addTransition(transition3)
        fsm.addTransition(transition4)

        while (true) {
            fsm.enterFrame(player)
            Thread.sleep(500)
        }
    }
}
